core/nats-macros/src/lib.rs

```diff
@@ -0,0 +1,67 @@
+use darling::{FromDeriveInput, FromMeta};
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{self, parse_macro_input, DeriveInput};
+
+#[derive(FromDeriveInput)]
+#[darling(attributes(typed_message))]
+struct Opts {
+    subject: NatsSubjectMacroInvocation,
+    #[darling(default)]
+    response: ResponseType,
+}
+
+#[derive(FromMeta, Clone, Debug)]
+struct ResponseType(syn::Type);
+
+impl Default for ResponseType {
+    fn default() -> Self {
+        ResponseType(syn::Type::from_string("NoReply").unwrap())
+    }
+}
+
+#[derive(Debug)]
+struct NatsSubjectMacroInvocation(syn::Macro);
+
+impl FromMeta for NatsSubjectMacroInvocation {
+    fn from_string(value: &str) -> darling::Result<Self> {
+        let props: Vec<&str> = value
+            .split('.')
+            .filter_map(|ea| ea.strip_prefix('#'))
+            .collect();
+        let mut fstring = value.to_string();
+        for prop in props.clone() {
+            fstring = fstring.replace(&("#".to_owned() + prop), "{}");
+        }
+        //note: this can likely be cleaned up with some judicious use of quote! and tokens!
+        let propstr = props
+            .iter()
+            .map(|prop| "self.".to_owned() + prop + ".as_subject_component()")
+            .reduce(|a, b| a + "," + &b)
```

Comment:
Use [join](https://doc.rust-lang.org/std/primitive.slice.html#method.join) for this.